primitives/arithmetic/src/fixed.rs

```diff
@@ -0,0 +1,1033 @@
+
+	/// Creates `self` from an integer number `int`.
+	/// 
+	/// Returns `None` if `int` exceeds accuracy.
+	fn checked_from_integer(int: Self::Inner) -> Option<Self>;
+
+	/// Creates `self` from a rational number. Equal to `n / d`.
+	///
+	/// Saturates if `d == 0` or `n / d` exceeds accuracy.
+	fn from_rational<N: FixedPointOperand, D: FixedPointOperand>(n: N, d: D) -> Self;
+
+	/// Creates `self` from a rational number. Equal to `n / d`.
+	///
+	/// Returns `None` if `d == 0` or `n / d` exceeds accuracy.
+	fn checked_from_rational<N: FixedPointOperand, D: FixedPointOperand>(n: N, d: D) -> Option<Self>;
+
+	/// Checked multiplication for integer type `N`.
+	fn checked_mul_int<N: FixedPointOperand>(self, other: N) -> Option<N>;
+
+	/// Saturating multiplication for integer type `N`.
+	fn saturating_mul_int<N: FixedPointOperand>(self, other: N) -> N;
+
+	/// Checked division for integer type `N`.
+	fn checked_div_int<N: FixedPointOperand>(self, other: N) -> Option<N>;
+
+	/// Saturating division for integer type `N`.
+	fn saturating_div_int<N: FixedPointOperand>(self, other: N) -> N;
+
+	/// Performs a saturated multiplication and accumulate by unsigned number.
+	///
+	/// Returns a saturated `int + (self * int)`.
+	fn saturating_mul_int_acc<N: FixedPointOperand>(self, int: N) -> N;
+
+	/// Saturating absolute value. Returning MAX if `parts == Inner::MIN` instead of overflowing.
+	fn saturating_abs(self) -> Self;
+
+	/// Takes the reciprocal (inverse), `1 / self`.
+	fn reciprocal(self) -> Option<Self> {
+		Self::one().checked_div(&self)
+	}
+
+	/// Returns zero.
+	fn zero() -> Self;
+
+	/// Checks if the number is zero.
+	fn is_zero(&self) -> bool;
+
+	/// Returns one.
+	fn one() -> Self;
+
+	/// Checks if the number is positive.
+	fn is_positive(self) -> bool;
+
+	/// Checks if the number is negative.
+	fn is_negative(self) -> bool;
+
+	/// Returns the integer part.
+	fn trunc(self) -> Self;
+
+	/// Returns the fractional part.
+	fn frac(self) -> Self;
+
+	/// Returns the smallest integer greater than or equal to a number.
+	fn ceil(self) -> Self;
+
+	/// Returns the largest integer less than or equal to a number.
+	fn floor(self) -> Self;
+
+	/// Returns the nearest integer to a number. Round half-way cases away from 0.0.
+	fn round(self) -> Self;
+
+	/// Computes the absolute value of self.
+	fn abs(self) -> Self;	
+}
+
+struct I129 {
+	value: u128,
+	negative: bool,
+}
+
+impl<N: FixedPointOperand> From<N> for I129 {
+	fn from(n: N) -> I129 {
+		if n < N::zero() {
+			let value: u128 = n.checked_neg()
+				.map(|n| n.unique_saturated_into())
+				.unwrap_or(N::max_value().unique_saturated_into().saturating_add(1));
+			I129 { value, negative: true }
+		} else {
+			I129 { value: n.unique_saturated_into(), negative: false }
+		}
+	}
+}
+
+fn from_i129<N: FixedPointOperand>(n: I129) -> Option<N> {
+	let max_plus_one: u128 = N::max_value().unique_saturated_into().saturating_add(1);
```

Comment:
How would `max_plus_one` and saturating add work here when N is u128?

Reply:
Good question! It should work fine when except when `n = (u128::max, true)`, in that case we don't know if the result should be `i128::min` or `i128::min + 1 (-u128::max)`. Mmm, I will add an `if` for that case (using `checked_add`) and return `i128::min + 1` which is (more) correct.

Reply:
Good point, probably I should separate the case when `n = (u128::max(), true)`, the other cases should work fine.